docs/adr/0014-auth-for-settings-redesign.md

```diff
@@ -0,0 +1,95 @@
+# Authentication in Settings Redesign
+
+- Status: proposed
+- Deciders: Les Orchard, Lauren Zugai, Ben Bangert
+- Date: 2020-04-15
+
+Issue: https://jira.mozilla.com/browse/FXA-1559
+
+## Context and Problem Statement
+
+The Settings Redesign app needs to query & mutate the same protected user data as the existing settings app hosted on content-server. This will require some form of authentication & authorization to manage that data.
+
+## Decision Drivers
+
+- Smooth UX
+- Security
+- Development velocity
+- Ease of integration
+
+## Considered Options
+
+- Reuse existing sessionToken on content-server with auth-server APIs
+- Authenticate via OAuth2 to use auth-server APIs
+- Authenticate via OAuth2 to use DB directly
+- Implement own login sessions
+
+## Decision Outcome
+
+Chosen option: "Reuse existing sessionToken on content-server with auth-server APIs", because it will have the least impact on existing auth-server implementation and infrastructure outside the Settings Redesign app in general.
+
+### Positive Consequences
+
+- The new Settings Redesign app can reuse existing auth-server APIs in the same way as existing settings code. This can constrain the novelty in the project to the redesigned & reimplemented settings UX.
+
+### Negative Consequences
+
+- We don't have an entirely greenfield project, which could be a bit of a drag. But, rediscovering lessons learned in existing code can also be a drag.
+
+## Pros and Cons of the Options
+
+### Reuse existing sessionToken on content-server with auth-server APIs
+
+The Settings Redesign frontend will be hosted at `accounts.firefox.com/beta/settings` per [FXA-1539](https://jira.mozilla.com/browse/FXA-1539) / [#4819](https://github.com/mozilla/fxa/issues/4819). This means it will have access to the same credentials as the original settings frontend on content-server.
+
+- Pros:
+  - Requires few or no auth-server changes - APIs already expect sessionToken
+  - Can be done with seamless UX - user visits new settings without confirmation
+- Neutral:
+  - GraphQL server needs to proxy auth-server APIs
```

Comment:
I don't have a lot of context on the intended architecture here, so it's not entirely clear to me what this means in practice. Will the new settings page talk to a new GraphQL server endpoint, which in turn talks to existing auth-server APIs? Or will the new settings page talk to a combination of existing auth-server APIs and the new GraphQL endpoint?

(I ask because it's not really possible to "proxy" sessionToken authentication in the same way that you might pass through an OAuth bearer token, thanks to the way we use Hawk for request signing).

Reply:
Ohhh I forgot about Hawk 😓  Yeah, I think the notion is that the new redesign frontend would talk to a GraphQL backend, which in turn would talk to auth-server APIs.

Reply:
We're not going to literally proxy the requests from the redesign frontend. The redesign frontend will make requests to the GQL service. In turn, the GQL service will orchestrate requests to auth-server APIs as necessary to fulfill its own queries & mutations.

But passing the client-side session token down to the server-side GQL service so it can make requests with Hawk basically defeats the purpose of Hawk, doesn't it?

Reply:
> (I ask because it's not really possible to "proxy" sessionToken authentication in the same way that you might pass through an OAuth bearer token, thanks to the way we use Hawk for request signing).

Welp, that pretty much puts the nail in this then. We'll have to use OAuth. Thanks for the input!

Reply:
> passing the client-side credentials down to the server-side GQL service so it can
> make its own requests with Hawk basically defeats the purpose of Hawk, doesn't it?

Aye, let's not do that.

You *could* arrange for some server-side secret to authenticate requests between the GQL service and the auth-server. So web content would make hawk-authenticated requests to the GQL, which is authenticates, and then makes its own signed request to the auth-server on the user's behalf. But even as I type that out it sounds unwieldy...

I do think you're in a good position to do a simpler version of what the payments-server does though. The settings page can use the `sessionToken` from localStorage to grant itself an OAuth access token with appropriate scopes, then use that OAuth access token for talking to the GQL server. IIUC that promises to be much simpler than the bouncing-back-and-forth-between-domains that the payments server needs to do to get an OAuth access token.